swift/Sources/FlatBuffers/Verifier.swift

```diff
@@ -0,0 +1,361 @@
+      throw FlatbuffersErrors.maximumTables
+    }
+    _depth += 1
+
+    if _depth > _options._maxDepth {
+      throw FlatbuffersErrors.maximumDepth
+    }
+
+    return TableVerifier(
+      position: position,
+      vtable: vtablePosition,
+      vtableLength: length,
+      verifier: &self)
+  }
+
+  ///
+  /// - Parameter position:
+  /// - Returns:
+  internal mutating func getValue<T>(at position: Int) throws -> T {
+    try inBuffer(position: position, of: T.self)
+    return _buffer.read(def: T.self, position: position)
+  }
+
+  internal mutating func derefOffset(position: Int) throws -> Int {
+    try inBuffer(position: position, of: Int32.self)
+
+    let offset = _buffer.read(def: Int32.self, position: position)
+    // switching to int32 since swift's default Int is int64
+    let _int32Position = Int32(position)
+
+    let reportedOverflow: (partialValue: Int32, overflow: Bool)
+    if offset > 0 {
+      reportedOverflow = _int32Position.subtractingReportingOverflow(abs(offset))
+    } else {
+      reportedOverflow = _int32Position.addingReportingOverflow(abs(offset))
+    }
+
+    /// since `subtractingReportingOverflow` & `addingReportingOverflow` returns true,
+    /// if there is overflow we return failure
+    if reportedOverflow.overflow || reportedOverflow.partialValue > _buffer.capacity {
+      throw FlatbuffersErrors.signedOffsetOutOfBounds(
+        offset: Int(offset),
+        position: position)
+    }
+
+    return Int(reportedOverflow.partialValue)
+  }
+}
+
+///
+public struct TableVerifier {
+
+  /// position of current table in `ByteBuffer`
+  fileprivate var _position: Int
+  fileprivate var _vtable: Int
+  fileprivate var _vtableLength: Int
+  fileprivate var _verifier: Verifier
+
+  ///
+  /// - Parameters:
+  ///   - position:
+  ///   - vtable:
+  ///   - vtableLength:
+  ///   - verifier:
+  internal init(
+    position: Int,
+    vtable: Int,
+    vtableLength: Int,
+    verifier: inout Verifier)
+  {
+    _position = position
+    _vtable = vtable
+    _vtableLength = vtableLength
+    _verifier = verifier
+  }
+
+  ///
+  /// - Parameter field:
+  /// - Throws:
+  /// - Returns:
+  internal mutating func dereference(_ field: VOffset) throws -> Int? {
+    if field >= _vtableLength {
+      return nil
+    }
+
+    let offset: VOffset = try _verifier.getValue(at: _vtable + Int(field))
+
+    if offset > 0 {
+      return _position + Int(offset)
+    }
+    return nil
+  }
+
+  ///
+  /// - Parameters:
+  ///   - field:
+  ///   - fieldName:
+  ///   - required:
+  ///   - type:
+  /// - Throws:
+  public mutating func visit<T>(
+    field: VOffset,
+    fieldName: String,
+    required: Bool,
+    type: T.Type) throws where T: Verifiable
+  {
+    let derefValue = try dereference(field)
+
+    if let value = derefValue {
+      try T.verify(&_verifier, at: value, of: T.self)
+      return
+    }
+    if required {
+      throw FlatbuffersErrors.requiredFieldDoesntExist(
+        position: field,
```

Comment:
Should we add do catch statements to propagate the error? since currently we just get an error like this; `.missalignedPointer(..)` but we don't know which table/value we referencing. Because swift will not throw an error with its caller stack.

Reply:
At least in other languages support for figuring out what field was corrupt and how is minimal, as it is typically quite meaningless (it can be the consequence of something that came before it). Corrupt is corrupt, and needs to be dealt with at the source.

Thus, I am not even a fan of exceptions in this case, I'd use a `bool` return like in C++.

Reply:
> Error handling in Swift resembles exception handling in other languages, with the use of the try, catch and throw keywords. Unlike exception handling in many languages—including Objective-C—error handling in Swift does not involve unwinding the call stack, a process that can be computationally expensive. As such, the performance characteristics of a throw statement are comparable to those of a return statement.

@aardappel According to the swift docs, there won't be a difference between throwing an error or returning a bool.

Reply:
@aardappel so what do you think regarding throws? since as i explained above its considered as a return statement in swift

Reply:
I guess that's fine, if it is common for Swift code to use this error handling instead of bools. Still, even if so, you typically have this kind of error handling (for example in Rust) when a function can both result in an expected result OR and error code. Here, we only have a single result: is this thing legal or not. A bool still seems better to me in that case. Failure to verify is not an "error" at the verifier level, it's what the verifier does.

But in the end, you know Swift best, so your call.

Reply:
Well, in swift most of the libraries use `throw` instead of result since the result code will just be a huge mess, example:
```swift
let ref1 = deref(offset)
let ref2 = deref(offset2)

guard let _ref1 = ref1.get() else { return .failure(ref1.error) }
guard let _ref2 = ref2.get() else { return .failure(ref2.error) }

/// more code with guards 
...
```
and so on, each time we will be calling the verifier or the generated code for the verifier would have a huge boiler plate of guard statements. whereas, `throw` will simply just return to the caller directly, instead of making sure that there is an error each time the code is doing something. 

Since unlike rust, an error will not be directly returing from the function @CasperN correct me if i am wrong
```rust
self.in_buffer::<u32>(pos)?; /// if this returns an error the rest of the function wont be called
/// terminated.
Ok(u32::from_le_bytes([
       self.buffer[pos],
       self.buffer[pos + 1],
       self.buffer[pos + 2],
       self.buffer[pos + 3],
    ]))
```
So a simple way to do it would be just `throwing` which will mimics the `result` behavior in rust.

Reply:
> At least in other languages support for figuring out what field was corrupt and how is minimal, as it is typically quite meaningless (it can be the consequence of something that came before it). Corrupt is corrupt, and needs to be dealt with at the source.

In Rust, I found the engineering effort for tracing exactly what was wrong and where to be useful, particularly for debugging. Also there was #6359 which was an unfortunate user facing bug that was trivial to diagnose due to the rich error message.

> Since unlike rust, an error will not be directly returing from the function @CasperN correct me if i am wrong

I can't tell exactly what you're asking, but i think the answer is "yes". That Rust code returns a `Result` type so yea its returns an error directly.The `?` is syntactic sugar for "if this expression is `Err`, return that error, otherwise continue", essentially equivalent to that `guard` boilerplate in swift.

Reply:
So we have two options now, whether we go with the current implementation with `throw` statements which are widely used in swift, and libraries example: [swift-protobuf](https://github.com/apple/swift-protobuf). Or use `result` with boilerplate in swift?

Reply:
I think it is up to you. Based only on what you say, `throw` sounds more idiomatic and thus it should be the preferred API.